lib/Sema/CSSimplify.cpp

```diff
@@ -1121,6 +1245,15 @@ class ArgumentFailureTracker : public MatchCallArgumentListener {
     return false;
   }
 
+  bool invalidArgumentsAfterCompletion() override {
+    assert(CS.isForCodeCompletion());
+    // This is only triggered when solving for code completion (which doesn't
+    // produce diagnostics) so no need to record a fix for this. Still increase
+    // the score to penalize this solution though.
+    CS.increaseScore(SK_Fix);
+    return false;
```

Comment:
I don't think `MatchCallArgumentListener` itself is passed in from anywhere (I think it's just a base class), but my understanding of it is that it's providing the default behaviour of disallowing all failures. I.e. if any kind of failure occurs, e.g. a label mismatch due to a typo, it fails fast and reports a mismatch, without giving any arg to param bindings. It basically answers "do these call arguments match this overload?" with a "yes + the determined arg/param bindings" or "no".

`ArgumentFailureTracker` tries to fix the issues reported to it if it can. So if there's a missing argument, for example, it records that fact (so we can provide diagnostics and fixits for it if this overload ends up being the "best" one), synthesizes an argument to stand in for it, and reports back to `matchCallArguments` to keep going. It answers "do these call arguments match this overload?" with a "yes + the determined arg/param bindings", "yes, if you perform these fixes + the determined arg/param bindings" or "no". We want this more lenient approach for code completion because a mismatch means there are no arg/param bindings and so no types for a code completion expression in call argument position. E.g in:
```
func foo(ccc: Int, y: SomeEnum) {}
foo(ccd: 2, y: .#^HERE^#) 
```
`ArgumentFailureTracker` provides a fix to relabel `ccd:` to `ccc:` so it still matches, we get param bindings, and can continue solving to produce a solution where the code completion expression's type is `SomeEnum`. If it didn't provide a fix and instead reported failure, we'd get no solutions for invalid cases like this.

Reply:
Because of the change I mentioned above to not remove the args after the completion position (and associated binding) I think I'll need to change this to record a fix anyway so I can avoid increasing the score twice if matchCallArguments reports invalid arguments after the completion but still binds one or more validly labelled post-completion args that end up having type mismatches later on.

Reply:
Thanks, that explanation makes sense to me. 👍